sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/WindowedSubscriber.java

```diff
@@ -0,0 +1,1157 @@
+            .addKeyValue(UPSTREAM_REQUESTED_KEY, requested)
+            .addKeyValue(DIFFERENCE_KEY, difference);
+
+        if (difference > 0) {
+            Operators.addCap(REQUESTED, this, difference);
+            logger.log("Initialized: request-upstream:true.");
+            s.request(difference);
+        } else {
+            logger.log("Initialized: request-upstream:false.");
+        }
+    }
+
+    /**
+     * Post a timed-out or canceled work to the drain-loop to process.
+     *
+     * @param w canceled or timed-out work.
+     */
+    private void postTimedOutOrCanceledWork(WindowWork<T> w) {
+        timedOutOrCanceledWorkQueue.add(w);
+        drain();
+    }
+
+    /**
+     * Gets the error describing the reason for subscriber (parent) termination.
+     *
+     * @return the subscriber (parent) termination reason.
+     */
+    private Throwable getTerminalError() {
+        assert isDoneOrCancelled();
+        if (done) {
+            final Throwable e = this.error;
+            return e != null ? e : new RuntimeException(String.format(terminatedMessage, "upstream-completion"));
+        }
+        return new RuntimeException(String.format(terminatedMessage, "downstream-cancel"));
+    }
+
+    /**
+     * Optional configurations for {@link WindowedSubscriber}.
+     *
+     * @param <T> the type of items in the windows that {@link WindowedSubscriber} produces.
+     */
+    static final class WindowedSubscriberOptions<T> {
+        private Consumer<T> releaser;
+        private Duration nextItemTimout;
+        private Function<Flux<T>, Flux<T>> windowDecorator;
+        private boolean cleanCloseStreamingWindowOnTerminate;
+
+        /**
+         * Creates the optional configurations for {@link WindowedSubscriber}.
+         */
+        WindowedSubscriberOptions() {
+            this.releaser = null;
+            this.nextItemTimout = null;
+            this.windowDecorator = null;
+            this.cleanCloseStreamingWindowOnTerminate = false;
+        }
+
+        /**
+         * Gets the callback to drop the upstream items when there is no active or pending window to deliver those.
+         *
+         * @return the callback to drop the items in the absence of a window.
+         */
+        private Consumer<T> getReleaser() {
+            return releaser;
+        }
+
+        /**
+         * Determines how long to wait for the next item in the window after the previous item. If the next item
+         * does not arrive within the wait time, then the window will be closed.
+         *
+         * @return the next window item timeout.
+         */
+        private Duration getNextItemTimout() {
+            return nextItemTimout;
+        }
+
+        /**
+         * Gets the decorator that add hooks to inspect the window events (items and termination).
+         *
+         * @return the window decorator.
+         */
+        private Function<Flux<T>, Flux<T>> getWindowDecorator() {
```

Comment:
I feel like I missed this in code somewhere... when is this used?

Reply:
An example usage looks like below (from SynchronousReceiver)

```java
    private WindowedSubscriber<ServiceBusReceivedMessage> createSubscriber() {
        ...

        final WindowedSubscriberOptions<ServiceBusReceivedMessage> options 
            = new WindowedSubscriberOptions<>();
        ...
        options.setWindowDecorator(this::traceDecorator); 
        ...
        return new WindowedSubscriber<>(
           Collections.singletonMap(ENTITY_PATH_KEY, entityPath), TERMINAL_MESSAGE, options);
    }

    private Flux<ServiceBusReceivedMessage> traceDecorator(Flux<ServiceBusReceivedMessage> toDecorate) {
        return tracer.traceSyncReceive(SYNC_RECEIVE_SPAN_NAME, toDecorate);
        ...
    }
```

Reply:
Ahh... for the tracer. I See, I See.